### PR TITLE
Adds count_on_hand to Spree::Variant to resolve n+1 queries

### PR DESCRIPTION
**Pull request by bryanmtl**
This is a proof of concept, an attempt to resolve the n+1 query problem introduced by the stock items setup and as described [here](https://github.com/spree/spree/issues/5971). This approach should maintain backwards compatibility for stores that don't track inventory. 

Further work should be done to optimize when the counter cache gets recomputed (ideally in the background).

### CODE REVIEW

**Comment by TeatroIO** [comment]
I've prepared a stage to preview changes. <a href='http://dynamomtl-feature-total-on-hand-optimizations.spree-spree-9ed724b77adec04c49c2.ttrcloud.com' target='_blank'>Open stage</a> or <a href='http://dynamomtl-feature-total-on-hand-optimizations.spree-spree-9ed724b77adec04c49c2.ttrcloud.com/teatro.log' target='_blank'>view logs</a>.

**Comment by gmacdougall** [comment]
:-1: to caching count on hand on products and/or variants. This isn't necessary to solve the problem that we're having. We have the ability to efficiently look up stock for a set of products using some intelligent queries. We're just not doing it.

**Comment by bryanmtl** [comment]
This isn't actually caching the count, it's just a workaround for the n+1 issue required to traverse the stock items.

File: core/db/migrate/20150220155706_add_total_on_hand_to_spree_products.rb [hunk 1]
```diff
@@ -0,0 +1,10 @@
+class AddTotalOnHandToSpreeProducts < ActiveRecord::Migration
+  def change
+    add_column :spree_products, :count_on_hand, :integer, null: false, default: 0
+  end
+
+  # seed the counts
+  Spree::Variant.all.each { |v| v.update_counter_cache }
```
Comment (BenMorganIO) [inline core/db/migrate/20150220155706_add_total_on_hand_to_spree_products.rb]: Please use `Spree::Variant.find_each(&:update_counter_cache)` instead.

File: core/db/migrate/20150220155706_add_total_on_hand_to_spree_products.rb [hunk 2]
```diff
@@ -0,0 +1,10 @@
+class AddTotalOnHandToSpreeProducts < ActiveRecord::Migration
+  def change
+    add_column :spree_products, :count_on_hand, :integer, null: false, default: 0
+    # seed the counts
+    Spree::Variant.all.each { |v| v.update_counter_cache }
```
Comment (BenMorganIO) [inline core/db/migrate/20150220155706_add_total_on_hand_to_spree_products.rb]: Some stores have variants into the hundreds of thousands. a) Use active job to update and b) use `#find_each`.

**Comment by BenMorganIO** [comment]
I know this is a work around, but I'm :+1: for a counter cache. I find them quite useful. I think stock_items already has a counter cache...

**Comment by gmacdougall** [comment]
I'm still :-1: for this. This data can be queried quickly and efficiently from our data store in the current setup. This is not how we should solve N+1 query issues. We should solve them by using better queries.

**Comment by bryanmtl** [comment]
@gmacdougall I tend to agree with you, but I was seeing better performance numbers with this approach. I'll post some benchmarks when I have some time over the weekend. I'm all for working on optimizing the queries with the goal of matching (or even exceeding) the performance gain I saw with this counter cache approach.

**Comment by gmacdougall** [comment]
I have a giant data set that I use for testing some stuff. (https://gist.github.com/gmacdougall/dbbe2cd68ff9d587c131)

It has 20 stock locations, 10,000 products, with between 1 and 10 variants for each product, and a random number of the stock locations having inventory.

Here's a simple query to get the count on hand for 50 products in the middle of the range and the PostgreSQL query plan for it:

```
spree_sandbox=# EXPLAIN ANALYZE select product.id, variant.id, SUM(count_on_hand)
  FROM spree_products product
  JOIN spree_variants variant ON (variant.product_id = product.id)
  JOIN spree_stock_items stock_item ON (stock_item.variant_id = variant.id)
  WHERE product.id IN (select id from spree_products WHERE id >= 3000 AND id < 3050)
  GROUP BY product.id, variant.id;
                                                                                     QUERY PLAN
------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 HashAggregate  (cost=1003.96..1071.55 rows=6759 width=12) (actual time=9.655..9.705 rows=345 loops=1)
   ->  Nested Loop  (cost=1.29..953.27 rows=6759 width=12) (actual time=0.122..6.367 rows=6900 loops=1)
         ->  Nested Loop  (cost=0.86..178.98 rows=338 width=8) (actual time=0.078..0.616 rows=345 loops=1)
               ->  Nested Loop  (cost=0.57..133.03 rows=50 width=8) (actual time=0.030..0.195 rows=50 loops=1)
                     ->  Index Only Scan using spree_products_pkey on spree_products  (cost=0.29..5.29 rows=50 width=4) (actual time=0.021..0.040 rows=50 loops=1)
                           Index Cond: ((id >= 3000) AND (id < 3050))
                           Heap Fetches: 0
                     ->  Index Only Scan using spree_products_pkey on spree_products product  (cost=0.29..2.55 rows=1 width=4) (actual time=0.002..0.002 rows=1 loops=50)
                           Index Cond: (id = spree_products.id)
                           Heap Fetches: 0
               ->  Index Scan using index_spree_variants_on_product_id on spree_variants variant  (cost=0.29..0.85 rows=7 width=8) (actual time=0.004..0.006 rows=7 loops=50)
                     Index Cond: (product_id = product.id)
         ->  Index Scan using index_spree_stock_items_on_variant_id on spree_stock_items stock_item  (cost=0.43..2.09 rows=20 width=8) (actual time=0.005..0.011 rows=20 loops=345)
               Index Cond: (variant_id = variant.id)
 Total runtime: 9.893 ms
(15 rows)
```

This counts the inventory for 345 variants in 20 stock location (6900 stock items) and does it in under 10ms on my i3-4130T, a perfectly reasonable amount of time. The performance issues here aren't related to our data structure, they're related to our code querying the data in an inefficient manner.

**Comment by bryanmtl** [comment]
Right and, to be clear, I never said this was an issue with the underlying data model. This was simply a proof of concept, a starting point to start easing some of the pressure we're seeing in production (on 2.3 and 2.4) when the product catalog is first loaded.

**Comment by gmacdougall** [comment]
I'm confused then. Do you want this to be merged, or development to continue along the path of adding count_on_hand to variant and/or product?

I think this change is bad, and should not be merged. I'm opposed to any duplication of data, and addition of callbacks which makes our models more complicated in a scenario where we can already get the data we want efficiently from our data model.

**Comment by bryanmtl** [comment]
Sorry @gmacdougall - the open PR is confusing. I don't intend to have this merged in like this. It's simply a proof of concept. If I get it to a point that I think makes sense and that doesn't require any significant data duplication, I'll bump the thread.

**Comment by gmacdougall** [comment]
Okay, that sounds good. I'm glad we're not on different pages.

**Comment by JDutil** [comment]
Going to close for now we can always reopen later.

**Pull request closed without merging**